tests/t0114_gateway_subdomains_test.go

```diff
@@ -2,61 +2,562 @@ package tests
+
+		// ## ============================================================================
+		// ## Test support for overriding implicit defaults
+		// ## ============================================================================
+		// TODO(lidel): We won't implement this in conformance testing.
+	}
+
+	if SubdomainGateway.IsEnabled() {
+		Run(t, tests)
+	}
+}
+
+func testGatewayWithManyProtocols(t *testing.T, label string, hint string, reqUrl interface{}, expected CResponse) []CTest {
+	t.Helper()
+
+	baseUrl := ""
+	baseReq := Request()
+
+	switch req := reqUrl.(type) {
+	case string:
+		baseUrl = reqUrl.(string)
+	case RequestBuilder:
+		baseReq = req
+		baseUrl = req.GetURL()
+	default:
+		t.Fatalf("invalid type for reqUrl: %T", reqUrl)
+	}
+
+	u, err := url.Parse(baseUrl)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Because you might be testing an IPFS node in CI, or on your local machine, the test are designed
+	// to test the subdomain behavior (querying http://{CID}.my-subdomain-gateway.io/) even if the node is
+	// actually living on http://127.0.0.1:8080 or somewhere else.
+	//
+	// The test knows two addresses:
+	// 		- GatewayUrl: the url we connect to, it might be "dweb.link", "127.0.0.1:8080", etc.
+	// 		- SubdomainGatewayUrl: the url we test for subdomain requests, it might be "dweb.link", "localhost", "example.com", etc.
+
+	// host is the hostname of the gateway we are testing, it might be `localhost` or `example.com`
+	host := u.Host
+
+	// raw url is the url but we replace the host with our local url, it might be `http://127.0.0.1/ipfs/something`
+	u.Host = GatewayHost
+	rawUrl := u.String()
+
+	return []CTest{
 		{
-			Name: "request for 127.0.0.1/ipfs/{CID} stays on path",
-			Request: CRequest{
-				Url: fmt.Sprintf("ipfs/%s", CIDv1),
-			},
-			Response: CResponse{
-				StatusCode: 200,
-				Body:       Contains(CID_VAL),
-			},
+			Name: fmt.Sprintf("%s (direct HTTP)", label),
+			Hint: fmt.Sprintf("%s\n%s", hint, "direct HTTP request (hostname in URL, raw IP in Host header)"),
+			Request: baseReq.
+				URL(rawUrl).
+				DoNotFollowRedirects().
+				Headers(
+					Header("Host", host),
+				).
+				Request(),
+			Response: expected,
 		},
 		{
-			Name: "request for localhost/ipfs/{CIDv1} returns HTTP 301 Moved Permanently",
-			Request: CRequest{
-				DoNotFollowRedirects: true,
-				Url:                  fmt.Sprintf("/ipfs/%s", CIDv1),
-			},
-			Response: CResponse{
-				StatusCode: 301,
-				Headers: map[string]interface{}{
-					"Location": Contains("http://%s.ipfs.localhost:8080", CIDv1),
-				},
-			},
+			Name: fmt.Sprintf("%s (HTTP proxy)", label),
+			Hint: fmt.Sprintf("%s\n%s", hint, "HTTP proxy (hostname is passed via URL)"),
+			Request: baseReq.
+				URL(baseUrl).
+				Proxy(GatewayUrl).
+				DoNotFollowRedirects().
+				Request(),
+			Response: expected,
+			// TODO(lidel): do we need to test different http behavior ?
+			// The previous test was using 2 proxy (--proxy and --proxy1.0)
```

Comment:
@lidel below, in [proxy.go](https://github.com/ipfs/gateway-conformance/pull/9/files#diff-df03cc864581cbd298d0dce0186ff5350c18c5733c46ccaaa32b929d38ab673eR1) we implement the proxy and tunnel proxy equivalents. The test used proxy1.0 before; ([ref](https://github.com/ipfs/kubo/blob/668d0b2fa4b43a1474ea0d1deab5295810a55f68/test/sharness/t0114-gateway-subdomains.sh#L56)) should we support HTTP 1.0, 1.1, and 2.0 explicitly for that first iteration?

Reply:
We only care about real world proxy modes, afaik only ones that matter are the three supported by `curl` (default vs HTTP proxy CONNECT 1.1 or 1.0)

- https://curl.se/docs/manpage.html#-x
- https://curl.se/docs/manpage.html#--proxy10  
- https://curl.se/docs/manpage.html#-p


I did not see anything specific to HTTP/2 in the wild, we don't need to worry about it yet.

Reply:
Thanks for the details! I looked into the docs and implementations,

**We don't test HTTP 1.0; there's a typo in sharness.**

Note the doc for proxytunnel:

> The only difference between this and the HTTP proxy option [-x, --proxy](https://curl.se/docs/manpage.html#-x), is that attempts to use CONNECT through the proxy will specify an HTTP 1.0 protocol instead of the default HTTP 1.1.

We don't use `--proxytunnel` with `--proxy1.0` which means that our tests [A](https://github.com/ipfs/kubo/blob/668d0b2fa4b43a1474ea0d1deab5295810a55f68/test/sharness/t0114-gateway-subdomains.sh#L56) and [B](https://github.com/ipfs/kubo/blob/668d0b2fa4b43a1474ea0d1deab5295810a55f68/test/sharness/t0114-gateway-subdomains.sh#L49) are testing the same thing.

Demo gist with every outputs: https://gist.github.com/laurentsenta/18b37a8405e4fee0d819102dde52e95b

**golang seems to ignore HTTP 1.0**

https://pkg.go.dev/net/http#Request

```
// For client requests, these fields are ignored. The HTTP
// client code always uses either HTTP/1.1 or HTTP/2.
// See the docs on Transport for details.
```

The test would be non-trivial to implement

**But**

We're not really exercising HTTP 1.0 anyway, we never keep the connection between requests, and we always pass a Host header. At best, we'd check that the gateway accepts our user agent.

My recommendation:
- we create an issue to add test HTTP 1.0 eventually but don't act on it for now.

Reply:
Nice catch! Sgtm. If nobody complained about lack of HTTP 1.0 here, means we can skip / remove.